### ISSUE TITLE
PHP 7.4 'ErrorException: Invalid characters passed for attempted conversion, these have been ignored'

### ISSUE THREAD

**Issue by ohnotnow** [barryvdh/laravel-dompdf#636]
Hi,

First of all - thanks for the package! :-)  I have a project I'm just testing with php 7.4.  When it tries to return the pdf it throws an exception :
```
$pdf = PDF::loadView('order.pdf', ['order' => $order]);

return $pdf->download($order->getPdfName());
```
```
ErrorException: Invalid characters passed for attempted conversion, these have been ignored

/Users/me/Documents/code/project/vendor/dompdf/dompdf/lib/Cpdf.php:2545
/Users/me/Documents/code/project/vendor/dompdf/dompdf/lib/Cpdf.php:2669
/Users/me/Documents/code/project/vendor/dompdf/dompdf/src/Adapter/CPDF.php:1006
/Users/me/Documents/code/project/vendor/dompdf/dompdf/src/FontMetrics.php:310
/Users/me/Documents/code/project/vendor/dompdf/dompdf/src/FrameReflower/Text.php:197
/Users/me/Documents/code/project/vendor/dompdf/dompdf/src/FrameReflower/Text.php:372
/Users/me/Documents/code/project/vendor/dompdf/dompdf/src/FrameDecorator/AbstractFrameDecorator.php:895
/Users/me/Documents/code/project/vendor/dompdf/dompdf/src/FrameReflower/Block.php:845
/Users/me/Documents/code/project/vendor/dompdf/dompdf/src/FrameDecorator/AbstractFrameDecorator.php:895
/Users/me/Documents/code/project/vendor/dompdf/dompdf/src/FrameReflower/Block.php:845
/Users/me/Documents/code/project/vendor/dompdf/dompdf/src/FrameDecorator/AbstractFrameDecorator.php:895
/Users/me/Documents/code/project/vendor/dompdf/dompdf/src/FrameReflower/Page.php:141
/Users/me/Documents/code/project/vendor/dompdf/dompdf/src/FrameDecorator/AbstractFrameDecorator.php:895
/Users/me/Documents/code/project/vendor/dompdf/dompdf/src/Dompdf.php:847
/Users/me/Documents/code/project/vendor/barryvdh/laravel-dompdf/src/PDF.php:195
/Users/me/Documents/code/project/vendor/barryvdh/laravel-dompdf/src/PDF.php:142
/Users/me/Documents/code/project/vendor/barryvdh/laravel-dompdf/src/PDF.php:165
```

If I make the blade template entirely empty it runs ok - but even if I add 'hello' it throws the exception.  Only seems to trigger on php 7.4.  This on Laravel 6.6.2, PHP 7.4.0, laravel-dompdf 0.8.5.

**Comment by neoteknic** [comment]
https://github.com/dompdf/dompdf/issues/2003
look my fix
@barryvdh Can you fork dompdf ? It's not maintained (or very slow) 2 months no update...
Just a quick fix dompdf\lib\Cpdf.php (this fix remove < php5.6 compat too)

` if ($c >= 0) {
                                if (ctype_xdigit($c) && ctype_xdigit($n)) {
                                    $data['codeToName'][$c] = $n;
                                }
                                $data['C'][$c] = $width;
                            } else {
                                $data['C'][$n] = $width;
                            }`

**Comment by neoteknic** [comment]
@ohnotnow fix here : 
https://github.com/dompdf/dompdf/pull/2027/files

**Comment by ohnotnow** [comment]
@neoteknic thanks for that - if I get desperate I'll pull that in :-)

**Comment by robertotcestari** [comment]
same problem here. Please merge this asap!

**Comment by pouriabagheri** [comment]
there is a recent fix for this bug on https://github.com/dompdf/dompdf/issues/2003

**Comment by bcmarcos03** [comment]
> [dompdf/dompdf#2003](https://github.com/dompdf/dompdf/issues/2003)
> look my fix
> @barryvdh Can you fork dompdf ? It's not maintained (or very slow) 2 months no update...
> Just a quick fix dompdf\lib\Cpdf.php (this fix remove < php5.6 compat too)
> 
> ` if ($c >= 0) { if (ctype_xdigit($c) && ctype_xdigit($n)) { $data['codeToName'][$c] = $n; } $data['C'][$c] = $width; } else { $data['C'][$n] = $width; }`

Hello, i tried this as i could see that it worked. But as it is in the vendor folder it won't get into my Github changes. How can I make this changes in my server using git or, what's the best practice in this cases? 

Thank you

**Comment by bkuhl** [comment]
The best practice approach would be to either wait for a 7.4 compatible release or fork the project with the fix you need.

**Comment by Michaeltlee** [comment]
`dompdf/dompdf` released version 0.8.4 that includes some changes so it will work on PHP 7.4. https://github.com/dompdf/dompdf/releases/tag/v0.8.4

**Comment by mahmudulhsn** [comment]
I have the same problem.... i am using 0.8.2 How can i fix it??
![image](https://user-images.githubusercontent.com/50780152/79068392-5ba4b800-7ce8-11ea-92e4-cb9bea6116a4.png)

**Comment by iscluismartinez101** [comment]
> [dompdf/dompdf#2003](https://github.com/dompdf/dompdf/issues/2003)
> look my fix
> @barryvdh Can you fork dompdf ? It's not maintained (or very slow) 2 months no update...
> Just a quick fix dompdf\lib\Cpdf.php (this fix remove < php5.6 compat too)
> 
> ` if ($c >= 0) { if (ctype_xdigit($c) && ctype_xdigit($n)) { $data['codeToName'][$c] = $n; } $data['C'][$c] = $width; } else { $data['C'][$n] = $width; }`

Muchas gracias, eres el mejor me tomo mucho tiempo tratar de solucionarlo y tu eres muy grande gracias !!!

**Comment by muhardiansah** [comment]
> there is a recent fix for this bug on [dompdf/dompdf#2003](https://github.com/dompdf/dompdf/issues/2003)

this works..

**Comment by OliverGrimsley** [comment]
@muhardiansah - the fix listed there will get the dompdf library to not crash, however at least for me (on the current release of PHP 7.4 in production, PHP 8 in dev) - my PDF is now gibberish - the image that I normally have at the top is placed at some random location above the margin and is cut off, and all the text is written on top of the existing text, so its unusable.  This was working fine as of 12/4, so something happened in an update either to PHP or this package since then.  Example - 

![Screen Shot 2020-12-15 at 6 40 08 AM](https://user-images.githubusercontent.com/12251451/102211101-210ba480-3ea1-11eb-8209-1956e5acadcf.jpg)

**Comment by Vicftz** [comment]
Hi @OliverGrimsley , I got the same problem! What version of dompdh are you using in dev ?

**Comment by OliverGrimsley** [comment]
> Hi @OliverGrimsley , I got the same problem! What version of dompdh are you using in dev ?

@Vicftz - This is v0.8.7 of this package, which requires ^0.8 of the underlying dompdf package, and which I have v0.8.3 of that dompdf package installed.   I spent a lot of time yesterday diagnosing - so I actually completely removed dompdf and this wrapper from composer, deleted my prior service provider, then re-installed to the most current version, so I was testing on the latest version.  I tried every set of solutions above and also in the dompdf issues (by manually changing my vendor file for dompdf - knowing it would require a fork if I fixed it) - no go - all the solutions fixed the crash, but none of them fixed the display issue.  I even ran my HTML through a validator, checked end tags - nothing worked.  I literally tried a piece of text with no HTML at all - it also re-arranged itself above the image.  Something in a recent update caused this, as this was working perfectly on 12/4.  I have not had the time to go back and check my git commits to see what I changed.  My server has not changed, so it must be an update to another package.

**Comment by Vicftz** [comment]
Ok 100% the same, i will take a look in my side, but if you tried every thing, I think we just need to wait for a patch from dompdf .... I keep you posted !
Thanks for your answer.

**Comment by ShenLoong99** [comment]
> I have the same problem.... i am using 0.8.2 How can i fix it?? Can any one please help??
> its urgent....
> 
> ![image](https://user-images.githubusercontent.com/50780152/79068392-5ba4b800-7ce8-11ea-92e4-cb9bea6116a4.png)

I am also facing the exact same error. Anyone has solution about this?

**Comment by OliverGrimsley** [comment]
You need to use dompdf > 0.8.4 (as of right now the current version is 0.8.6).  If you are still running php 7.x it will work (at least, this error goes away) if you can upgrade to the most recent version.  However, dompdf itself is still not tagged for php 8 and hence at least for me in my development environment, I cannot get to the current version.  Even if I could, something is seriously off on the formatting as I noted above - and that issue persists for me even if I was able to upgrade to a current version.  I have had to work around it manually for now and probably going to move to something else.

**Comment by deanzod** [comment]
> You need to use dompdf > 0.8.4.
> 
> For me, I had to remove my dompdf dependency in my composer file, and then I removed this package and reinstalled, that bumped this package to .9, and then it upgraded dompdf and everything is working again.

Yes I had same issue...updating the package didn't work but when I ran `composer remove barryvdh/laravel-dompdf` then `composer require barryvdh/laravel-dompdf` it seemed to update whatever it needed to update with the fresh install and all working now

**Comment by waltlin1** [comment]
After searching the web for a solution, I finally found a workaround for this issue. I first tried to modify the function, but it came out even worse. I inserted the following code directly above line 2546, where in my case the error happened:

`error_reporting(E_ALL ^ E_DEPRECATED);`

<img width="763" alt="Bildschirmfoto 2021-05-12 um 13 52 56" src="https://user-images.githubusercontent.com/51401541/117970792-5cfc4600-b329-11eb-9a5f-cdc4cef3f94e.png">

For me that worked without a rollback to php 7.3 or lower. I'm running PHP 7.4
